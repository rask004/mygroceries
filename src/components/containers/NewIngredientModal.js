import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {FaSave, FaRegTimesCircle} from 'react-icons/fa';
import TextField from '@material-ui/core/TextField';
import Grid from '@material-ui/core/Grid';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import DialogTitle from '@material-ui/core/DialogTitle';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import Typography from '@material-ui/core/Typography';
import {assign, omit, uniq} from 'lodash';
import 'typeface-roboto';
import '../../css/recipemodals.css';


const errorMessages = {
    quantity: "quantity requires a decimal or whole number greater than 0.",
    name: "ingredient name must have a value.",
    brand: "brand must have a value.",
    category: "category must have a value.",
    subcategory: "subcategory must have a value.",
    coo: "country of origin must have a value.",
    defaultunit: "default unit must have a value.",
    unit: "unit must have a value.",
};


const initialModalState = {
    name: "",
    category: "",
    subcategory: "",
    coo: "",
    brand: "",
    defaultunit: "",
    quantity: 0,
    unit: "",
    errors: {
        name: true,
        category: true,
        subcategory: true,
        coo: true,
        brand: true,
        defaultunit: true,
        quantity: true,
        unit: true,
        message: errorMessages.quantity,
    }
};


const makeDatalist = (products, key) => {
    return uniq(products.map(x => x[key])).map( (x, i) => (
        <option key={i} value={x} />
    ));
};


class NewIngredientModal extends Component {
    constructor(props) {
        super(props);
        this.state = initialModalState;

        this.handleChange = this.handleChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
        this.handleClose = this.handleClose.bind(this);
        this.handleErrorChecking = this.handleErrorChecking.bind(this);
    }

    handleChange(event) {
        const target = event.target;
        const name = target.name;
        const value = target.value;

        const newState = {[name]: value, unit: this.state.unit, errors: this.state.errors};
        if(name === "defaultunit") {
            newState.unit = value;
            newState.errors.unit = false;
        }
        this.setState(newState);
        this.handleErrorChecking(event);
    }

    handleErrorChecking(event) {
        const target = event.target;
        const name = target.name;
        const value = target.value;
        const errors = this.state.errors;

        if(value === "" || (name === "quantity" && (isNaN(value) || Number(value) <= 0))) {
            errors[name] = true;
        } else if (name !== undefined) {
            errors[name] = false;
        }

        errors.message = "";
        for(let errorName in errors) {
            if(errorName !== "message" && errors[errorName]) {
                errors.message = errorMessages[errorName];
                break;
            }
        }
        
        this.setState({errors: errors});
    }

    handleSubmit(event) {
        event.preventDefault();
        // I don't know why it's putting "" in errors state, but this fixes it. 
        const errorState = omit(this.state.errors, [""]);

        for(let name in errorState) {
            if(name === "message") {
                continue;
            }
            const isError = errorState[name];
            if (isError) {
                return false;
            }
        }

        const ingredient = omit(assign({}, this.state), ["quantity", "unit", "errors"]);

        const quantity = this.state.quantity;
        const unit = this.state.unit;
        const ingredientAdded = this.props.onAddIngredient(ingredient, quantity, unit);
        if (ingredientAdded) {
            this.setState(initialModalState);
        }
    }

    handleClose() {
        this.setState(initialModalState);
        this.props.onClose();
    }

    render() {

        const products = this.props.products;

        const names = makeDatalist(products, "name");
        const countries = makeDatalist(products, "coo");
        const categories = makeDatalist(products, "category");
        const subcategories = makeDatalist(products, "subcategory");
        const defaultUnits = makeDatalist(products, "defaultUnit");
        const brands = makeDatalist(products, "brand");

        return (
            <Dialog open={this.props.show} onClose={this.handleClose}
                maxWidth="md" fullWidth scroll="body">
                <form className="add-ingredient" 
                    onSubmit={this.handleSubmit}>

                    <datalist id="names">
                        {names}
                    </datalist>
                    <datalist id="categories">
                        {categories}
                    </datalist>
                    <datalist id="subcategories">
                        {subcategories}
                    </datalist>
                    <datalist id="brands">
                        {brands}
                    </datalist>
                    <datalist id="countries">
                        {countries}
                    </datalist>
                    <datalist id="default-units">
                        {defaultUnits}
                    </datalist>

                    <DialogTitle disableTypography className="title">
                        <Typography variant="display2">
                            Add Ingredient
                        </Typography>
                    </DialogTitle>
                    <DialogContent>
                        <Grid container className="row">
                            <Grid item xs={12} md={5}>
                                <TextField required name="name"
                                    placeholder="Name of Ingredient" 
                                    label="Name"
                                    variant="outlined"
                                    fullWidth
                                    margin="normal"
                                    InputLabelProps={{
                                        shrink: true,
                                    }}
                                    onChange={this.handleChange}
                                    value={this.state.name} 
                                    inputProps={{
                                        list: "names"
                                    }}
                                    error={this.state.errors.name}
                                />
                            </Grid>
                            <Grid item xs={false} md={7}></Grid>
                        </Grid>

                        <Grid container className="row">
                            <Grid item xs={12} md={5}>
                                <TextField required name="category"
                                    placeholder="Category" 
                                    label="Category"
                                    variant="outlined"
                                    fullWidth
                                    margin="normal"
                                    InputLabelProps={{
                                        shrink: true,
                                    }}
                                    onChange={this.handleChange}
                                    value={this.state.category} 
                                    inputProps={{
                                        list: "categories"
                                    }}
                                    error={this.state.errors.category}
                                />
                            </Grid>

                            <Grid item xs={false} md={2}></Grid>

                            <Grid item xs={12} md={5}>
                                <TextField required name="subcategory"
                                    label="Subcategory"
                                    placeholder="SubCategory"
                                    variant="outlined"
                                    fullWidth
                                    margin="normal"
                                    InputLabelProps={{
                                        shrink: true,
                                    }}
                                    onChange={this.handleChange}
                                    value={this.state.subcategory} 
                                    inputProps={{
                                        list: "subcategories"
                                    }}
                                    error={this.state.errors.subcategory}
                                />
                            </Grid>
                        </Grid>

                        <Grid container className="row">
                            <Grid item xs={12} md={5}>
                                <TextField required name="brand"
                                    label="Brand"
                                    placeholder="Brand"
                                    variant="outlined"
                                    fullWidth
                                    margin="normal"
                                    InputLabelProps={{
                                        shrink: true,
                                    }}
                                    onChange={this.handleChange}
                                    value={this.state.brand} 
                                    inputProps={{
                                        list: "brands"
                                    }}
                                    error={this.state.errors.brand}
                                />
                            </Grid>
                            
                            <Grid item xs={false} md={2}>    
                            </Grid>

                            <Grid item xs={12} md={5}>
                                <TextField required name="coo"
                                    placeholder="Country of Origin" 
                                    label="Country"
                                    variant="outlined"
                                    fullWidth
                                    margin="normal"
                                    InputLabelProps={{
                                        shrink: true,
                                    }}
                                    onChange={this.handleChange}
                                    value={this.state.coo} 
                                    inputProps={{
                                        list: "countries"
                                    }}
                                    error={this.state.errors.coo}
                                />
                            </Grid>
                        </Grid>

                        <Grid container className="row">
                            <Grid item xs={false} md={7}>    
                            </Grid>
                            
                            <Grid item xs={12} md={5}>
                                <TextField required name="defaultunit"
                                    placeholder="Default Unit (singular)" 
                                    label="Default Unit"
                                    variant="outlined"
                                    fullWidth
                                    margin="normal"
                                    InputLabelProps={{
                                        shrink: true,
                                    }}
                                    onChange={this.handleChange}
                                    value={this.state.defaultunit} 
                                    inputProps={{
                                        list: "default-units"
                                    }}
                                    error={this.state.errors.defaultunit}
                                />
                            </Grid>
                        </Grid>

                        <Grid container className="row">
                            <Grid item xs={12} md={12}>
                            </Grid>
                        </Grid>

                        <Grid container className="row">

                            <Grid item xs={false} md={2}>  
                            </Grid>

                            <Grid item xs={12} md={3}>
                                <TextField required name="quantity"
                                    label="Quantity"
                                    variant="outlined"
                                    fullWidth
                                    placeholder="Quantity" 
                                    onChange={this.handleChange}
                                    value={this.state.quantity} 
                                    margin="normal"
                                    InputLabelProps={{
                                        shrink: true,
                                    }}
                                    error={this.state.errors.quantity}
                                />
                            </Grid>
                            
                            <Grid item xs={false} md={2}>    
                            </Grid>

                            <Grid item xs={12} md={5}>
                                <TextField required name="unit"
                                    label="Unit"
                                    variant="outlined"
                                    fullWidth
                                    placeholder="Unit (singular)"
                                    onChange={this.handleChange}
                                    value={this.state.unit}
                                    margin="normal"
                                    InputLabelProps={{
                                        shrink: true,
                                    }}
                                    inputProps={{
                                        list: "default-units"
                                    }}
                                    error={this.state.errors.unit}
                                />
                            </Grid>
                        </Grid>
                    </DialogContent>
                    <DialogActions className="button-pane">
                        <Button type="submit" variant="contained" color="primary" size="large"
                            onClick={this.handleErrorChecking}>
                            <FaSave/>
                        </Button>
                        <Typography color="error" align="center" inline className="error-message">
                            {this.state.errors.message}
                        </Typography>
                        <Button type="button" variant="contained" color="primary" size="large" 
                            onClick={this.handleClose}>
                            <FaRegTimesCircle/>
                        </Button>
                    </DialogActions>
                </form>
            </Dialog>
        );
    }
}

NewIngredientModal.propTypes = {
    products: PropTypes.array.isRequired,
    show: PropTypes.bool.isRequired,
    onClose: PropTypes.func.isRequired,
    onAddIngredient: PropTypes.func.isRequired
}


const mapStateToProps = state => {
	return {
        products: state.products
    }
}

export {NewIngredientModal};

export default connect(mapStateToProps)(NewIngredientModal);
