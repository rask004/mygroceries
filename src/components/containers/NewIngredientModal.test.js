import React from 'react';
import {create} from 'react-test-renderer';
import {render} from 'react-testing-library';

import {NewIngredientModal} from './NewIngredientModal';

import sampleData from '../../data/initialState.json';


//jest.mock('@material-ui/core/Dialog');


const products = sampleData.products;

const modal = <NewIngredientModal 
        products={products}
        show={true}
        onClose={jest.fn()}
        onAddIngredient={jest.fn()}
    />


describe('NewIngredientModal Component render tests', () => {

    test('renders correctly', () => {
        const component = create(modal);
        expect(component.toJSON()).toMatchSnapshot();
    });
});


describe('NewIngredientModal Component behaviour tests', () => {

    test('modal UI changes', () => {

        const component = create(modal).root;
        const nameField = component.findByProps({name:"name"})



        // check with different values inserted

        // check how state of relevant inputs changes

        // check warning message at bottom
        
        // incomplete
        expect("incomplete").not.toBeTruthy();
    });



    test('save button click', () => {

        // click button, check no save call

        // put valid values in inputs one by one, each time clicking button, checking no save call.

        // put final valid value in, check click makes correct call.
        
        // incomplete
        expect("incomplete").not.toBeTruthy();
    });

    test('cancel button click', () => {

        // 
        
        // incomplete
        expect("incomplete").not.toBeTruthy();
    });
});
