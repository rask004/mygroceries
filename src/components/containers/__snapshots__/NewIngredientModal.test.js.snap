// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewIngredientModal Component render tests renders correctly 1`] = `
<div>
  <form
    className="add-ingredient"
    onSubmit={[Function]}
  >
    <datalist
      id="names"
    >
      <option
        value="apple"
      />
      <option
        value="orange"
      />
      <option
        value="banana"
      />
      <option
        value="bluetop milk"
      />
      <option
        value="gherkin"
      />
      <option
        value="anaroba extra-strong coffee"
      />
    </datalist>
    <datalist
      id="categories"
    >
      <option
        value="produce"
      />
      <option
        value="foodstuffs"
      />
      <option
        value="canned foods"
      />
      <option
        value="drinks"
      />
    </datalist>
    <datalist
      id="subcategories"
    >
      <option
        value="fruit"
      />
      <option
        value="dairy"
      />
      <option
        value="pickles"
      />
      <option
        value="coffee"
      />
    </datalist>
    <datalist
      id="brands"
    >
      <option
        value="Pure Orchards LTD"
      />
      <option
        value="Dole LTD"
      />
      <option
        value="Anchor LTD"
      />
      <option
        value="San Fernando Produce"
      />
      <option
        value="Sunkiss Products"
      />
      <option
        value="Rio de Feliz Comidas"
      />
    </datalist>
    <datalist
      id="countries"
    >
      <option
        value="New Zealand"
      />
      <option
        value="Mexico"
      />
      <option
        value="California"
      />
      <option
        value="Australia"
      />
      <option
        value="Argentina"
      />
    </datalist>
    <datalist
      id="default-units"
    >
      <option
        value="each"
      />
      <option
        value="litre"
      />
      <option
        value="grams"
      />
    </datalist>
    <div
      className="MuiDialogTitle-root-1 title"
    >
      <h1
        className="MuiTypography-root-2 MuiTypography-display2-5"
      >
        Add Ingredient
      </h1>
    </div>
    <div
      className="MuiDialogContent-root-38"
    >
      <div
        className="MuiGrid-container-39 row"
      >
        <div
          className="MuiGrid-item-40 MuiGrid-grid-xs-12-79 MuiGrid-grid-md-5-100"
        >
          <div
            className="MuiFormControl-root-136 MuiFormControl-marginNormal-137 MuiFormControl-fullWidth-139"
          >
            <label
              className="MuiFormLabel-root-151 MuiFormLabel-error-154 MuiInputLabel-error-143 MuiFormLabel-required-156 MuiInputLabel-required-144 MuiInputLabel-root-140 MuiInputLabel-formControl-145 MuiInputLabel-animated-148 MuiInputLabel-shrink-147 MuiInputLabel-outlined-150"
              data-shrink={true}
            >
              Name
              <span
                className="MuiFormLabel-asterisk-157 MuiFormLabel-error-154 MuiInputLabel-error-143"
              >
                 *
              </span>
            </label>
            <div
              className="MuiInputBase-root-171 MuiOutlinedInput-root-158 MuiInputBase-error-177 MuiOutlinedInput-error-163 MuiInputBase-fullWidth-180 MuiInputBase-formControl-172"
              onClick={[Function]}
            >
              <fieldset
                aria-hidden={true}
                className="MuiPrivateNotchedOutline-root-188 MuiOutlinedInput-notchedOutline-165"
                style={
                  Object {
                    "paddingLeft": 8,
                  }
                }
              >
                <legend
                  className="MuiPrivateNotchedOutline-legend-189"
                  style={
                    Object {
                      "width": 0,
                    }
                  }
                >
                  <span
                    dangerouslySetInnerHTML={
                      Object {
                        "__html": "&#8203;",
                      }
                    }
                  />
                </legend>
              </fieldset>
              <input
                aria-invalid={true}
                className="MuiInputBase-input-181 MuiOutlinedInput-input-166"
                disabled={false}
                list="names"
                name="name"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="Name of Ingredient"
                required={true}
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          className="MuiGrid-item-40 MuiGrid-grid-md-7-102"
        />
      </div>
      <div
        className="MuiGrid-container-39 row"
      >
        <div
          className="MuiGrid-item-40 MuiGrid-grid-xs-12-79 MuiGrid-grid-md-5-100"
        >
          <div
            className="MuiFormControl-root-136 MuiFormControl-marginNormal-137 MuiFormControl-fullWidth-139"
          >
            <label
              className="MuiFormLabel-root-151 MuiFormLabel-error-154 MuiInputLabel-error-143 MuiFormLabel-required-156 MuiInputLabel-required-144 MuiInputLabel-root-140 MuiInputLabel-formControl-145 MuiInputLabel-animated-148 MuiInputLabel-shrink-147 MuiInputLabel-outlined-150"
              data-shrink={true}
            >
              Category
              <span
                className="MuiFormLabel-asterisk-157 MuiFormLabel-error-154 MuiInputLabel-error-143"
              >
                 *
              </span>
            </label>
            <div
              className="MuiInputBase-root-171 MuiOutlinedInput-root-158 MuiInputBase-error-177 MuiOutlinedInput-error-163 MuiInputBase-fullWidth-180 MuiInputBase-formControl-172"
              onClick={[Function]}
            >
              <fieldset
                aria-hidden={true}
                className="MuiPrivateNotchedOutline-root-188 MuiOutlinedInput-notchedOutline-165"
                style={
                  Object {
                    "paddingLeft": 8,
                  }
                }
              >
                <legend
                  className="MuiPrivateNotchedOutline-legend-189"
                  style={
                    Object {
                      "width": 0,
                    }
                  }
                >
                  <span
                    dangerouslySetInnerHTML={
                      Object {
                        "__html": "&#8203;",
                      }
                    }
                  />
                </legend>
              </fieldset>
              <input
                aria-invalid={true}
                className="MuiInputBase-input-181 MuiOutlinedInput-input-166"
                disabled={false}
                list="categories"
                name="category"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="Category"
                required={true}
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          className="MuiGrid-item-40 MuiGrid-grid-md-2-97"
        />
        <div
          className="MuiGrid-item-40 MuiGrid-grid-xs-12-79 MuiGrid-grid-md-5-100"
        >
          <div
            className="MuiFormControl-root-136 MuiFormControl-marginNormal-137 MuiFormControl-fullWidth-139"
          >
            <label
              className="MuiFormLabel-root-151 MuiFormLabel-error-154 MuiInputLabel-error-143 MuiFormLabel-required-156 MuiInputLabel-required-144 MuiInputLabel-root-140 MuiInputLabel-formControl-145 MuiInputLabel-animated-148 MuiInputLabel-shrink-147 MuiInputLabel-outlined-150"
              data-shrink={true}
            >
              Subcategory
              <span
                className="MuiFormLabel-asterisk-157 MuiFormLabel-error-154 MuiInputLabel-error-143"
              >
                 *
              </span>
            </label>
            <div
              className="MuiInputBase-root-171 MuiOutlinedInput-root-158 MuiInputBase-error-177 MuiOutlinedInput-error-163 MuiInputBase-fullWidth-180 MuiInputBase-formControl-172"
              onClick={[Function]}
            >
              <fieldset
                aria-hidden={true}
                className="MuiPrivateNotchedOutline-root-188 MuiOutlinedInput-notchedOutline-165"
                style={
                  Object {
                    "paddingLeft": 8,
                  }
                }
              >
                <legend
                  className="MuiPrivateNotchedOutline-legend-189"
                  style={
                    Object {
                      "width": 0,
                    }
                  }
                >
                  <span
                    dangerouslySetInnerHTML={
                      Object {
                        "__html": "&#8203;",
                      }
                    }
                  />
                </legend>
              </fieldset>
              <input
                aria-invalid={true}
                className="MuiInputBase-input-181 MuiOutlinedInput-input-166"
                disabled={false}
                list="subcategories"
                name="subcategory"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="SubCategory"
                required={true}
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-container-39 row"
      >
        <div
          className="MuiGrid-item-40 MuiGrid-grid-xs-12-79 MuiGrid-grid-md-5-100"
        >
          <div
            className="MuiFormControl-root-136 MuiFormControl-marginNormal-137 MuiFormControl-fullWidth-139"
          >
            <label
              className="MuiFormLabel-root-151 MuiFormLabel-error-154 MuiInputLabel-error-143 MuiFormLabel-required-156 MuiInputLabel-required-144 MuiInputLabel-root-140 MuiInputLabel-formControl-145 MuiInputLabel-animated-148 MuiInputLabel-shrink-147 MuiInputLabel-outlined-150"
              data-shrink={true}
            >
              Brand
              <span
                className="MuiFormLabel-asterisk-157 MuiFormLabel-error-154 MuiInputLabel-error-143"
              >
                 *
              </span>
            </label>
            <div
              className="MuiInputBase-root-171 MuiOutlinedInput-root-158 MuiInputBase-error-177 MuiOutlinedInput-error-163 MuiInputBase-fullWidth-180 MuiInputBase-formControl-172"
              onClick={[Function]}
            >
              <fieldset
                aria-hidden={true}
                className="MuiPrivateNotchedOutline-root-188 MuiOutlinedInput-notchedOutline-165"
                style={
                  Object {
                    "paddingLeft": 8,
                  }
                }
              >
                <legend
                  className="MuiPrivateNotchedOutline-legend-189"
                  style={
                    Object {
                      "width": 0,
                    }
                  }
                >
                  <span
                    dangerouslySetInnerHTML={
                      Object {
                        "__html": "&#8203;",
                      }
                    }
                  />
                </legend>
              </fieldset>
              <input
                aria-invalid={true}
                className="MuiInputBase-input-181 MuiOutlinedInput-input-166"
                disabled={false}
                list="brands"
                name="brand"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="Brand"
                required={true}
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          className="MuiGrid-item-40 MuiGrid-grid-md-2-97"
        />
        <div
          className="MuiGrid-item-40 MuiGrid-grid-xs-12-79 MuiGrid-grid-md-5-100"
        >
          <div
            className="MuiFormControl-root-136 MuiFormControl-marginNormal-137 MuiFormControl-fullWidth-139"
          >
            <label
              className="MuiFormLabel-root-151 MuiFormLabel-error-154 MuiInputLabel-error-143 MuiFormLabel-required-156 MuiInputLabel-required-144 MuiInputLabel-root-140 MuiInputLabel-formControl-145 MuiInputLabel-animated-148 MuiInputLabel-shrink-147 MuiInputLabel-outlined-150"
              data-shrink={true}
            >
              Country
              <span
                className="MuiFormLabel-asterisk-157 MuiFormLabel-error-154 MuiInputLabel-error-143"
              >
                 *
              </span>
            </label>
            <div
              className="MuiInputBase-root-171 MuiOutlinedInput-root-158 MuiInputBase-error-177 MuiOutlinedInput-error-163 MuiInputBase-fullWidth-180 MuiInputBase-formControl-172"
              onClick={[Function]}
            >
              <fieldset
                aria-hidden={true}
                className="MuiPrivateNotchedOutline-root-188 MuiOutlinedInput-notchedOutline-165"
                style={
                  Object {
                    "paddingLeft": 8,
                  }
                }
              >
                <legend
                  className="MuiPrivateNotchedOutline-legend-189"
                  style={
                    Object {
                      "width": 0,
                    }
                  }
                >
                  <span
                    dangerouslySetInnerHTML={
                      Object {
                        "__html": "&#8203;",
                      }
                    }
                  />
                </legend>
              </fieldset>
              <input
                aria-invalid={true}
                className="MuiInputBase-input-181 MuiOutlinedInput-input-166"
                disabled={false}
                list="countries"
                name="coo"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="Country of Origin"
                required={true}
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-container-39 row"
      >
        <div
          className="MuiGrid-item-40 MuiGrid-grid-md-7-102"
        />
        <div
          className="MuiGrid-item-40 MuiGrid-grid-xs-12-79 MuiGrid-grid-md-5-100"
        >
          <div
            className="MuiFormControl-root-136 MuiFormControl-marginNormal-137 MuiFormControl-fullWidth-139"
          >
            <label
              className="MuiFormLabel-root-151 MuiFormLabel-error-154 MuiInputLabel-error-143 MuiFormLabel-required-156 MuiInputLabel-required-144 MuiInputLabel-root-140 MuiInputLabel-formControl-145 MuiInputLabel-animated-148 MuiInputLabel-shrink-147 MuiInputLabel-outlined-150"
              data-shrink={true}
            >
              Default Unit
              <span
                className="MuiFormLabel-asterisk-157 MuiFormLabel-error-154 MuiInputLabel-error-143"
              >
                 *
              </span>
            </label>
            <div
              className="MuiInputBase-root-171 MuiOutlinedInput-root-158 MuiInputBase-error-177 MuiOutlinedInput-error-163 MuiInputBase-fullWidth-180 MuiInputBase-formControl-172"
              onClick={[Function]}
            >
              <fieldset
                aria-hidden={true}
                className="MuiPrivateNotchedOutline-root-188 MuiOutlinedInput-notchedOutline-165"
                style={
                  Object {
                    "paddingLeft": 8,
                  }
                }
              >
                <legend
                  className="MuiPrivateNotchedOutline-legend-189"
                  style={
                    Object {
                      "width": 0,
                    }
                  }
                >
                  <span
                    dangerouslySetInnerHTML={
                      Object {
                        "__html": "&#8203;",
                      }
                    }
                  />
                </legend>
              </fieldset>
              <input
                aria-invalid={true}
                className="MuiInputBase-input-181 MuiOutlinedInput-input-166"
                disabled={false}
                list="default-units"
                name="defaultunit"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="Default Unit (singular)"
                required={true}
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-container-39 row"
      >
        <div
          className="MuiGrid-item-40 MuiGrid-grid-xs-12-79 MuiGrid-grid-md-12-107"
        />
      </div>
      <div
        className="MuiGrid-container-39 row"
      >
        <div
          className="MuiGrid-item-40 MuiGrid-grid-md-2-97"
        />
        <div
          className="MuiGrid-item-40 MuiGrid-grid-xs-12-79 MuiGrid-grid-md-3-98"
        >
          <div
            className="MuiFormControl-root-136 MuiFormControl-marginNormal-137 MuiFormControl-fullWidth-139"
          >
            <label
              className="MuiFormLabel-root-151 MuiFormLabel-error-154 MuiInputLabel-error-143 MuiFormLabel-filled-155 MuiFormLabel-required-156 MuiInputLabel-required-144 MuiInputLabel-root-140 MuiInputLabel-formControl-145 MuiInputLabel-animated-148 MuiInputLabel-shrink-147 MuiInputLabel-outlined-150"
              data-shrink={true}
            >
              Quantity
              <span
                className="MuiFormLabel-asterisk-157 MuiFormLabel-error-154 MuiInputLabel-error-143"
              >
                 *
              </span>
            </label>
            <div
              className="MuiInputBase-root-171 MuiOutlinedInput-root-158 MuiInputBase-error-177 MuiOutlinedInput-error-163 MuiInputBase-fullWidth-180 MuiInputBase-formControl-172"
              onClick={[Function]}
            >
              <fieldset
                aria-hidden={true}
                className="MuiPrivateNotchedOutline-root-188 MuiOutlinedInput-notchedOutline-165"
                style={
                  Object {
                    "paddingLeft": 8,
                  }
                }
              >
                <legend
                  className="MuiPrivateNotchedOutline-legend-189"
                  style={
                    Object {
                      "width": 0,
                    }
                  }
                >
                  <span
                    dangerouslySetInnerHTML={
                      Object {
                        "__html": "&#8203;",
                      }
                    }
                  />
                </legend>
              </fieldset>
              <input
                aria-invalid={true}
                className="MuiInputBase-input-181 MuiOutlinedInput-input-166"
                disabled={false}
                name="quantity"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="Quantity"
                required={true}
                type="text"
                value={0}
              />
            </div>
          </div>
        </div>
        <div
          className="MuiGrid-item-40 MuiGrid-grid-md-2-97"
        />
        <div
          className="MuiGrid-item-40 MuiGrid-grid-xs-12-79 MuiGrid-grid-md-5-100"
        >
          <div
            className="MuiFormControl-root-136 MuiFormControl-marginNormal-137 MuiFormControl-fullWidth-139"
          >
            <label
              className="MuiFormLabel-root-151 MuiFormLabel-error-154 MuiInputLabel-error-143 MuiFormLabel-required-156 MuiInputLabel-required-144 MuiInputLabel-root-140 MuiInputLabel-formControl-145 MuiInputLabel-animated-148 MuiInputLabel-shrink-147 MuiInputLabel-outlined-150"
              data-shrink={true}
            >
              Unit
              <span
                className="MuiFormLabel-asterisk-157 MuiFormLabel-error-154 MuiInputLabel-error-143"
              >
                 *
              </span>
            </label>
            <div
              className="MuiInputBase-root-171 MuiOutlinedInput-root-158 MuiInputBase-error-177 MuiOutlinedInput-error-163 MuiInputBase-fullWidth-180 MuiInputBase-formControl-172"
              onClick={[Function]}
            >
              <fieldset
                aria-hidden={true}
                className="MuiPrivateNotchedOutline-root-188 MuiOutlinedInput-notchedOutline-165"
                style={
                  Object {
                    "paddingLeft": 8,
                  }
                }
              >
                <legend
                  className="MuiPrivateNotchedOutline-legend-189"
                  style={
                    Object {
                      "width": 0,
                    }
                  }
                >
                  <span
                    dangerouslySetInnerHTML={
                      Object {
                        "__html": "&#8203;",
                      }
                    }
                  />
                </legend>
              </fieldset>
              <input
                aria-invalid={true}
                className="MuiInputBase-input-181 MuiOutlinedInput-input-166"
                disabled={false}
                list="default-units"
                name="unit"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="Unit (singular)"
                required={true}
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="button-pane"
    >
      <button
        className="MuiButtonBase-root-216 MuiButton-root-190 MuiButton-contained-201 MuiButton-containedPrimary-202 MuiButton-raised-204 MuiButton-raisedPrimary-205 MuiButton-sizeLarge-214"
        disabled={false}
        onBlur={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="submit"
      >
        <span
          className="MuiButton-label-191"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M433.941 129.941l-83.882-83.882A48 48 0 0 0 316.118 32H48C21.49 32 0 53.49 0 80v352c0 26.51 21.49 48 48 48h352c26.51 0 48-21.49 48-48V163.882a48 48 0 0 0-14.059-33.941zM224 416c-35.346 0-64-28.654-64-64 0-35.346 28.654-64 64-64s64 28.654 64 64c0 35.346-28.654 64-64 64zm96-304.52V212c0 6.627-5.373 12-12 12H76c-6.627 0-12-5.373-12-12V108c0-6.627 5.373-12 12-12h228.52c3.183 0 6.235 1.264 8.485 3.515l3.48 3.48A11.996 11.996 0 0 1 320 111.48z"
            />
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root-219"
        />
      </button>
      <p
        className="MuiTypography-root-2 MuiTypography-body1-11 MuiTypography-colorError-36 MuiTypography-alignCenter-25 MuiTypography-inline-37 error-message"
      >
        quantity requires a decimal or whole number greater than 0.
      </p>
      <button
        className="MuiButtonBase-root-216 MuiButton-root-190 MuiButton-contained-201 MuiButton-containedPrimary-202 MuiButton-raised-204 MuiButton-raisedPrimary-205 MuiButton-sizeLarge-214 action-cancel"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiButton-label-191"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm0 448c-110.5 0-200-89.5-200-200S145.5 56 256 56s200 89.5 200 200-89.5 200-200 200zm101.8-262.2L295.6 256l62.2 62.2c4.7 4.7 4.7 12.3 0 17l-22.6 22.6c-4.7 4.7-12.3 4.7-17 0L256 295.6l-62.2 62.2c-4.7 4.7-12.3 4.7-17 0l-22.6-22.6c-4.7-4.7-4.7-12.3 0-17l62.2-62.2-62.2-62.2c-4.7-4.7-4.7-12.3 0-17l22.6-22.6c4.7-4.7 12.3-4.7 17 0l62.2 62.2 62.2-62.2c4.7-4.7 12.3-4.7 17 0l22.6 22.6c4.7 4.7 4.7 12.3 0 17z"
            />
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root-219"
        />
      </button>
    </div>
  </form>
</div>
`;
